Pass context to own commands and profile runner

creativeprojects opened the pull request:
- introduction of a `Context` struct to pass context information around commands and profiles.
- own commands can return a specific error to exit main with an error code (will be used by `run-schedule`)
- extract bits of code into methods to simplify `main`
- add a `pre` field to `ownCommand` so commands can check the context before running

That will simplify future work like #259 and #146

jkellerer reviewed (commented):
Like the change. Just have a few comments on the structs.

creativeprojects commented on context.go:
```diff
@@ -0,0 +1,44 @@
+	profile       *config.Profile
+	schedule      *config.Schedule // when profile is running with run-schedule command
+	sigChan       chan os.Signal
+}
```
neat 👍🏻

creativeprojects commented:
I've implemented your comments (just because it made sense 😆 )

I've also added a method `loadContext` which... I'm not too sure about it.
The problem is that we need to load many things before we can setup the logger. In case of `run-schedule` command we need to access `schedule-log`.

So that's why I moved the log target to the `loadContext` as well 😐

jkellerer commented:
I’ll also think about the log part and cross check it with the other PR.

jkellerer commented:
Setup logging in context creation actually makes sense since the log target can change (however can't seem to find it in this PR).

The only other concern I have is that all the `With...` funcs should return a copy of the context so that side effects are avoided (and that all of the funcs behave the same). Copy is cheap and easy anyway.

creativeprojects commented:
> Setup logging in context creation actually makes sense since the log target can change (however can't seem to find it in this PR).

It's creating `logTarget` in the `loadContext` method. The only choice at that point is from global or flags.

The rest is on the following PR actually. I thought that was enough changes for this one already:

To set the log from schedule, it's part of the `pre` callback (of `run-schedule` command)

https://github.com/creativeprojects/resticprofile/blob/run-schedule/commands.go#L614

jkellerer commented:
LGTM

creativeprojects commented:
Thanks 😉
